extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/security/RbacServiceSecurityTest.java

```diff
@@ -0,0 +1,84 @@
+package io.quarkus.websockets.next.test.security;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.net.URI;
+import java.util.Set;
+
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.security.ForbiddenException;
+import io.quarkus.security.test.utils.TestIdentityController;
+import io.quarkus.security.test.utils.TestIdentityProvider;
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.test.common.http.TestHTTPResource;
+import io.quarkus.websockets.next.OnError;
+import io.quarkus.websockets.next.OnTextMessage;
+import io.quarkus.websockets.next.WebSocket;
+import io.quarkus.websockets.next.test.utils.WSClient;
+import io.vertx.core.Vertx;
+
+public class RbacServiceSecurityTest extends SecurityTestBase {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot(root -> root.addClasses(Endpoint.class, AdminService.class, UserService.class,
+                    TestIdentityProvider.class, TestIdentityController.class, WSClient.class));
+
+    @Inject
+    Vertx vertx;
+
+    @TestHTTPResource("end")
+    URI endUri;
+
+    @BeforeAll
+    public static void setupUsers() {
+        TestIdentityController.resetRoles()
+                .add("admin", "admin", "admin")
+                .add("user", "user", "user");
+    }
+
+    @Test
+    public void testEndpoint() {
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(basicAuth("admin", "admin"), endUri);
+            client.sendAndAwait("hello"); // admin service
+            client.sendAndAwait("hi"); // forbidden
+            client.waitForMessages(2);
+            assertEquals(Set.of("24", "forbidden"), Set.copyOf(client.getMessages().stream().map(Object::toString).toList()));
+        }
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(basicAuth("user", "user"), endUri);
+            client.sendAndAwait("hello"); // forbidden
+            client.sendAndAwait("hi"); // user service
+            client.waitForMessages(2);
+            assertEquals(Set.of("42", "forbidden"), Set.copyOf(client.getMessages().stream().map(Object::toString).toList()));
```

Comment:
> Just so that people won't forget, it doesn't make too much of a sense to me to keep this open if you cannot change the identity.

Well, it depends on how the endpoint is used. You cannot change the identity but the message that resulted in a `ForbiddenException` could have been a mistake. It's very similar to a webapp where a user intentionally or unintentionally attemps to access a resource that is forbidden for this particular user :shrug:. Do we log off the user automatically?

> @mkouba speaking of what, according to the https://quarkus.io/guides/websockets-next-reference#websocket-endpoint-methods the `@OnError` is not required, am I correct? 

You're right.

> If so, what will happen when there is no `@OnError`, can we just close the connection in the future (not in this PR)?

Currently, we only log an error message. But we can certainly close the connection by default.

Reply:
So what about in situations where the endpoint itself has the security annotations? If the endpoint is @RolesAllowed("admin") and someone connects with a user role then ideally we would just reject the connection outright (at least, as an end user that would be the behavior I am expecting). 

I agree that it does not make sense to close the connection when invoking business methods that fail the permission check.

Reply:
> 
> Currently, we only log an error message. But we can certainly close the connection by default.

Is this something that an unauthenticated user can easily trigger? If you can easily trigger excessive logging on the server side it can be seen as a DOS vector.

Reply:
> So what about in situations where the endpoint itself has the security annotations? If the endpoint is @RolesAllowed("admin") and someone connects with a user role then ideally we would just reject the connection outright (at least, as an end user that would be the behavior I am expecting).

Hm, for this kind of stuff we would need an SPI that would allow us to perform a security check for a class/set of annotation instances - the current SPI seems to be very method-oriented. Actually, we could use the security checks like `io.quarkus.security.runtime.interceptor.check.RolesAllowedCheck` directly, or even mimic the logic used there - it's pretty straightforward. But ideally, the `SecurityCheckStorage` would declare something like `getSecurityCheck(Class<?>)`...

CC @michalvavrik

> > Currently, we only log an error message. But we can certainly close the connection by default.
> 
> Is this something that an unauthenticated user can easily trigger? If you can easily trigger excessive logging on the server side it can be seen as a DOS vector.

Yes, so we should disable excessive logging for security exceptions by default.

Reply:
Hm, the problem is that JAX-RS endpoints work diferently - each method handles an HTTP request, so it makes sense to map the security checks to methods. In our case, an endpoint method does not represent an HTTP request, in fact the endpoint instance is created and callbacks are invoked after the initial HTTP request upgrade finished successfully.

Maybe we should merge this PR and improve the integration in follow-up PRs. The discussion is getting harder to follow.

Reply:
Thanks Michal. Re closed connection - yes, I think that we could make it configurable and enable it by default, i.e. a connection is closed automatically if a security failure occurs.

Reply:
Are we sure logging is the best option when an unhandled failure of any kind occurs? Unlike HTTP there is no standard way to report back to the client that there was a problem, so closing the connection seems like a reasonable repose.

That said I think we should make sure that if the endpoint has security annotations on the class itself it should affect the upgrade process, and make sure it will send back a challenge if the user is not authenticated. Opening the connection and then immediately closing it or just essentially ignoring all messages are both the wrong things to do.